susemanager-utils/susemanager-sls/src/modules/ssh_agent.py

```diff
@@ -0,0 +1,84 @@
+import logging
+from salt.exceptions import CommandExecutionError
+try:
+    from salt.utils.path import which_bin as _which_bin
+except ImportError:
+    from salt.utils import which_bin as _which_bin
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'ssh_agent'
+
+__start_agent = '/usr/bin/ssh-agent'
+__ssh_add = '/usr/bin/ssh-add {}'
+
+def __virtual__():
+    '''
+    This module is always enabled while 'ssh-agent' is available.
+    '''
+    return __virtualname__ if _which_bin(['ssh-agent']) else (False, 'ssh-agent is not available')
+
+# def __call_ssh_tool(ssh_tool, cmd_args, **kwargs):
+#     log.debug("Calling ssh-agent: 'skuba {}' - Timeout: {}".format(cmd_args, timeout))
+#     try:
+#         ssh_tool_proc = salt.utils.timed_subprocess.TimedProc(
+#             [ssh_tool] + cmd_args.split(),
+#             stdout=subprocess.PIPE,
+#             stderr=subprocess.PIPE,
+#             timeout=timeout
+#         )
+#         ssh_tool_proc.run()
+#     except Exception as exc:
+#         error_msg = "Unexpected error while calling {}: {}".format(ssh_tool, exc)
+#         log.error(error_msg)
+#         raise CommandExecutionError(error_msg)    
+
+#     if ssh_tool_proc.process.returncode != 0 or ssh_tool_proc.stderr:
+#         error_msg = "Unexpected error {} when calling {}: {}".format(
+#                 ssh_tool_proc.returncode,
+#                 ssh_tool,
+#                 salt.utils.stringutils.to_str(ssh_tool_proc.stderr))
+#         log.error(error_msg)
+#         raise CommandExecutionError(error_msg)
+
+
+def start_agent(**kwargs):
+    # TODO add socket param
+    result = __salt__['cmd.run_all'](__start_agent, output_loglevel='quiet')
+    if result['retcode'] != 0:
+        raise CommandExecutionError(result['stderr'])
+
+    ssh_agent_lines = result['stdout'].splitlines()
+
+    variables = dict()
+    for line in ssh_agent_lines:
+        if line.startswith('SSH'):
+            var, *rest = line.split(';')
+            key, val = var.strip().split("=", 1)
+            variables[key] = val
+
+    __salt__['environ.setenv'](variables)
+    return variables
+
+
+def list_keys(**kwargs):
+    # TODO add socket param
+    result = __salt__['cmd.run_all'](__ssh_add.format('-l'), output_loglevel='quiet')
+    if result['retcode'] != 0:
+        raise CommandExecutionError(result['stderr'])
+    return result['stdout']
+
+
+def add_key(ssh_key_file, **kwargs):
+    # TODO add socket param
+    result = __salt__['cmd.run_all'](__ssh_add.format(ssh_key_file), output_loglevel='quiet')
```

Comment:
Ok, let me elaborate the explanation here. There are few reason why this should be avoided:

- Using `cmd.run` or `cmd.run_all` like this means, it also executes a shell to run the process (not really needed)
- Running a shell means we need to take care of possible code injection.
- In this particular case here, we don't perform any safety check for the `ssh_key_file` argument of the function (provided by the user), and we use that to feed the `cmd.run_all` argument. This currently allow extra command injection:

```
# salt-call --local cmd.run_all "date; echo my injected code"
local:
    ----------
    pid:
        15767
    retcode:
        0
    stderr:
    stdout:
        Mon Jun  1 12:16:30 WEST 2020
        my injected code
```

Therefore doing something like:
```
# salt-call --local ssh_agent.add_key "my_test_file; some malicious rm code"
```
would execute the malicious code.

To avoid this we could add `python_shell=False` to the `cmd.run` command or maybe using the unimplemented `__call_ssh_tool` method